torch/distributed/_tensor/placement_types.py

```diff
@@ -54,40 +54,70 @@ def _split_tensor(
+        ]
+        # Compute pad size on each chunk
+        pad_sizes = [
+            full_chunk_size - chunk_size for idx, chunk_size in enumerate(chunk_sizes)
+        ]
+
+        # Reuse tensor to fill empty chunk with empty tensor
+        num_empty_tensors = num_chunks - len(tensor_list)
+        tensor_size = list(tensor_list[0].size())
+        tensor_size = [
+            size if idx != self.dim else 0 for idx, size in enumerate(tensor_size)
+        ]
+        tensor = tensor.new_zeros(tensor_size)
+        for _ in range(num_empty_tensors):
+            tensor_list.append(tensor)
+
         if with_padding or contiguous:
             shard_list = []
-            for i, shard in enumerate(tensor_list):
-                if with_padding and idx_start_to_pad != 0 and i >= idx_start_to_pad:
-                    shard = self._pad_tensor(shard)
-                # input tensors are expected to be contiguous by the collective backend
+            for shard, pad_size in zip(tensor_list, pad_sizes):
+                # Fill the empty tensor with zeroes with padding.
+                if with_padding and pad_size > 0:
+                    shard = self._pad_tensor(shard, pad_size)
                 shard = shard.contiguous() if contiguous else shard
                 shard_list.append(shard)
-            return shard_list, idx_start_to_pad
+            return shard_list, pad_sizes
         else:
-            return tensor_list, idx_start_to_pad
+            return tensor_list, pad_sizes
 
-    def _pad_tensor(self, tensor: torch.Tensor) -> torch.Tensor:
-        # pad tensor by 1 on the shard dim
+    def _pad_tensor(
+        self,
+        tensor: torch.Tensor,
+        pad_size: int,
+    ) -> torch.Tensor:
         pad = [0, 0] * (tensor.ndim - self.dim)
-        pad[-1] = 1
+        pad[-1] = pad_size
         return torch.nn.functional.pad(tensor, pad)
 
-    def _unpad_tensor(self, tensor: torch.Tensor) -> torch.Tensor:
-        # unpad tensor by 1 on the shard dim
-        return tensor.narrow(self.dim, start=0, length=tensor.size(self.dim) - 1)
-
-    def _unpad_concat_tensor(
-        self, tensor: torch.Tensor, padding_idx: int, shard_count: int
+    def _unpad_tensor(
+        self,
+        tensor: torch.Tensor,
+        pad_size: int,
     ) -> torch.Tensor:
-        gathered_list = torch.chunk(tensor, shard_count, dim=self.dim)
-        gathered_list = [
-            self._unpad_tensor(gathered_tensor) if i >= padding_idx else gathered_tensor
-            for i, gathered_tensor in enumerate(gathered_list)
-        ]
-        return torch.cat(gathered_list, dim=self.dim)
+        tensor = tensor.narrow(
+            self.dim,
+            start=0,
+            length=tensor.size(self.dim) - pad_size,
+        )
+
+        # Explicitly return an empty tensor. Otherwise, even if the
+        # tensor is empty, the size won't be 0.
+        if tensor.numel() == 0:
+            return tensor.new_zeros([0])
```

Comment:
> Hmmm could you explain why we need to explicitly return an empty tensor? iiuc after `tensor.narrow` there would be some ranks have `tensor` with shape ([0, 8]) for example, that's still considered as empty tensor with no data, so that should work for us.

The reason is that we have test to compare an unpad tensor with original tensor_to_split.

https://github.com/pytorch/pytorch/blob/ada67c9d8a3ac61fa0af5b8a186d5ecb31765af5/test/distributed/_tensor/test_device_mesh.py#L211

In this case, the size of the two would fail the assert as one could be ([0, 8]) and the other one is ([0]), although both are just empty tensors([]). I guess I could update the test cases to make _unpad() more consistent.